### ISSUE TITLE
Look into speedup

### ISSUE THREAD

**Issue by arendMoerman** [PyPO-dev/PyPO#175]
CUDA has a lot of optimised intrinsic device functions (see [here](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__SINGLE.html#group__CUDA__MATH__INTRINSIC__SINGLE) for an overview of single-precision intrinsics), such as vector normalisation or squaring. Also, it seems that a lot of information in a block (e.g., source surface coordinates, source currents) is re-used internally, but still loaded/accessed by each thread separately. Instead, it might be better to load chunks into shared memory (noting the 48 kB shared memory limit per block), perform computations on that, accumulate results into (coalesced) global memory, and refresh chunk/repeat.

To summarise, I think we can get some speedup if we:
* Look into these intrinsics (for example, vector normalisation)
* Look into utilising more shared memory

Maybe an idea to look into this.

On another note, I think all global memory accesses (at least in the kernel) are coalesced, so thats good. Not sure about device functions though.
